File: 5_WebApps/ShopERP/Services/RealtimeDashboardService.cs
using VanAn.Shared.Domain;
using Microsoft.AspNetCore.SignalR;
using Microsoft.Extensions.Caching.Memory;

namespace VanAn.ShopERP.Services
{
    public interface IRealtimeDashboardService
    {
        Task<OrderMetrics> GetCurrentMetricsAsync();
        Task<List<Order>> GetRecentOrdersAsync(int count = 10);
        Task<List<Order>> GetActiveOrdersAsync();
        Task<bool> BroadcastOrderUpdateAsync(Order order);
        Task<bool> BroadcastMetricsUpdateAsync();
        Task<bool> SubscribeToUpdatesAsync(string connectionId);
        Task<bool> UnsubscribeFromUpdatesAsync(string connectionId);
        Task<List<DashboardAlert>> GetActiveAlertsAsync();
        Task<bool> CreateAlertAsync(DashboardAlert alert);
    }

    public class DashboardAlert
    {
        public string Id { get; set; } = Guid.NewGuid().ToString();
        public AlertType Type { get; set; }
        public string Title { get; set; } = string.Empty;
        public string Message { get; set; } = string.Empty;
        public AlertSeverity Severity { get; set; }
        public DateTime CreatedAt { get; set; }
        public DateTime? ExpiresAt { get; set; }
        public bool IsRead { get; set; }
        public string? ActionUrl { get; set; }
        public Dictionary<string, object> Metadata { get; set; } = [];
    }

    public enum AlertType
    {
        OrderCreated,
        OrderUpdated,
        OrderCancelled,
        HighValueOrder,
        SystemError,
        LowInventory,
        StaffAssignment
    }

    public enum AlertSeverity
    {
        Info,
        Warning,
        Error,
        Critical
    }

    public class RealtimeDashboardService(
        IOrderManagementService orderManagementService,
        IHubContext<DashboardHub> hubContext,
        ILogger<RealtimeDashboardService> logger,
        IMemoryCache cache) : IRealtimeDashboardService
    {
        private readonly IOrderManagementService _orderManagementService = orderManagementService;
        private readonly IHubContext<DashboardHub> _hubContext = hubContext;
        private readonly ILogger<RealtimeDashboardService> _logger = logger;
        private readonly IMemoryCache _cache = cache;
        private readonly List<string> _subscribedConnections = [];
        private readonly object _lock = new();

        public async Task<OrderMetrics> GetCurrentMetricsAsync()
        {
            try
            {
                string cacheKey = "dashboard_metrics";
                if (_cache.TryGetValue(cacheKey, out OrderMetrics? cachedMetrics))
                {
                    return cachedMetrics ?? new OrderMetrics();
                }

                OrderMetrics metrics = await _orderManagementService.GetOrderMetricsAsync();

                // Cache for 30 seconds
                _cache.Set(cacheKey, metrics, TimeSpan.FromSeconds(30));

                return metrics;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to get current dashboard metrics");
                return new OrderMetrics();
            }
        }

        public async Task<List<Order>> GetRecentOrdersAsync(int count = 10)
        {
            try
            {
                string cacheKey = $"recent_orders_{count}";
                if (_cache.TryGetValue(cacheKey, out List<Order>? cachedOrders))
                {
                    return cachedOrders ?? [];
                }

                List<Order> allOrders = await _orderManagementService.GetOrdersAsync();
                List<Order> recentOrders = allOrders
                    .OrderByDescending(o => o.CreatedAt)
                    .Take(count)
                    .ToList();

                // Cache for 15 seconds
                _cache.Set(cacheKey, recentOrders, TimeSpan.FromSeconds(15));

                return recentOrders;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to get recent orders");
                return [];
            }
        }

        public async Task<List<Order>> GetActiveOrdersAsync()
        {
            try
            {
                string cacheKey = "active_orders";
                if (_cache.TryGetValue(cacheKey, out List<Order>? cachedOrders))
                {
                    return cachedOrders ?? [];
                }

                List<Order> pendingOrders = await _orderManagementService.GetOrdersAsync(OrderStatusId.Pending);
                List<Order> processingOrders = await _orderManagementService.GetOrdersAsync(OrderStatusId.Processing);

                List<Order> activeOrders =
                [
                    .. pendingOrders.Concat(processingOrders)
                                        .OrderByDescending(o => o.CreatedAt)
,
                ];

                // Cache for 10 seconds
                _cache.Set(cacheKey, activeOrders, TimeSpan.FromSeconds(10));

                return activeOrders;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to get active orders");
                return [];
            }
        }

        public async Task<bool> BroadcastOrderUpdateAsync(Order order)
        {
            try
            {
                lock (_lock)
                {
                    if (_subscribedConnections.Count == 0)
                    {
                        return false; // No subscribers
                    }
                }

                // Invalidate relevant caches
                _cache.Remove("dashboard_metrics");
                _cache.Remove("recent_orders_10");
                _cache.Remove("active_orders");

                // Create alert for high-value orders
                if (order.TotalAmount > 1000000) // 1 million VND
                {
                    await CreateAlertAsync(new DashboardAlert
                    {
                        Type = AlertType.HighValueOrder,
                        Title = "High Value Order",
                        Message = $"Order {order.Id} with total {order.TotalAmount:N0} VND",
                        Severity = AlertSeverity.Info,
                        CreatedAt = DateTime.UtcNow,
                        ExpiresAt = DateTime.UtcNow.AddHours(1),
                        Metadata = new Dictionary<string, object>
                        {
                            ["OrderId"] = order.Id,
                            ["TotalAmount"] = order.TotalAmount
                        }
                    });
                }

                // Broadcast to all subscribed clients
                await _hubContext.Clients.All.SendAsync("OrderUpdated", new
                {
                    OrderId = order.Id,
                    Status = order.Status.ToString(),
                    order.TotalAmount,
                    order.CreatedAt,
                    order.UpdatedAt,
                    ItemCount = order.Items.Count
                });

                // Also broadcast updated metrics
                await BroadcastMetricsUpdateAsync();

                _logger.LogInformation("Broadcasted order update: {OrderId}", order.Id);
                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to broadcast order update: {OrderId}", order.Id);
                return false;
            }
        }

        public async Task<bool> BroadcastMetricsUpdateAsync()
        {
            try
            {
                lock (_lock)
                {
                    if (_subscribedConnections.Count == 0)
                    {
                        return false; // No subscribers
                    }
                }

                OrderMetrics metrics = await GetCurrentMetricsAsync();

                await _hubContext.Clients.All.SendAsync("MetricsUpdated", new
                {
                    metrics.TotalOrders,
                    metrics.PendingOrders,
                    metrics.ProcessingOrders,
                    metrics.CompletedOrders,
                    metrics.CancelledOrders,
                    metrics.TotalRevenue,
                    metrics.AverageOrderValue,
                    metrics.OrdersPerHour,
                    metrics.RevenuePerHour,
                    metrics.StatusBreakdown,
                    Timestamp = DateTime.UtcNow
                });

                _logger.LogInformation("Broadcasted metrics update");
                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to broadcast metrics update");
                return false;
            }
        }

        public async Task<bool> SubscribeToUpdatesAsync(string connectionId)
        {
            try
            {
                lock (_lock)
                {
                    if (!_subscribedConnections.Contains(connectionId))
                    {
                        _subscribedConnections.Add(connectionId);
                    }
                }

                // Send initial data
                await _hubContext.Clients.Client(connectionId).SendAsync("InitialData", new
                {
                    Metrics = await GetCurrentMetricsAsync(),
                    RecentOrders = await GetRecentOrdersAsync(10),
                    ActiveOrders = await GetActiveOrdersAsync(),
                    Alerts = await GetActiveAlertsAsync()
                });

                _logger.LogInformation("Client subscribed to dashboard updates: {ConnectionId}", connectionId);
                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to subscribe client to updates: {ConnectionId}", connectionId);
                return false;
            }
        }

        public async Task<bool> UnsubscribeFromUpdatesAsync(string connectionId)
        {
            try
            {
                lock (_lock)
                {
                    _subscribedConnections.Remove(connectionId);
                }

                _logger.LogInformation("Client unsubscribed from dashboard updates: {ConnectionId}", connectionId);
                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to unsubscribe client from updates: {ConnectionId}", connectionId);
                return false;
            }
        }

        public async Task<List<DashboardAlert>> GetActiveAlertsAsync()
        {
            try
            {
                string cacheKey = "active_alerts";
                if (_cache.TryGetValue(cacheKey, out List<DashboardAlert>? cachedAlerts))
                {
                    return cachedAlerts ?? [];
                }

                // In a real implementation, this would fetch from a database
                // For now, we'll return some sample alerts
                List<DashboardAlert> alerts =
                [
                    new DashboardAlert
                    {
                        Type = AlertType.SystemError,
                        Title = "System Performance",
                        Message = "Dashboard running normally",
                        Severity = AlertSeverity.Info,
                        CreatedAt = DateTime.UtcNow.AddMinutes(-5),
                        IsRead = false
                    }
                ];

                // Cache for 1 minute
                _cache.Set(cacheKey, alerts, TimeSpan.FromMinutes(1));

                return alerts;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to get active alerts");
                return [];
            }
        }

        public async Task<bool> CreateAlertAsync(DashboardAlert alert)
        {
            try
            {
                alert.CreatedAt = DateTime.UtcNow;

                // Invalidate alerts cache
                _cache.Remove("active_alerts");

                // Broadcast alert to all subscribed clients
                lock (_lock)
                {
                    if (_subscribedConnections.Count != 0)
                    {
                        _hubContext.Clients.All.SendAsync("AlertCreated", new
                        {
                            alert.Id,
                            alert.Type,
                            alert.Title,
                            alert.Message,
                            alert.Severity,
                            alert.CreatedAt,
                            alert.ExpiresAt,
                            alert.ActionUrl,
                            alert.Metadata
                        });
                    }
                }

                _logger.LogInformation("Created dashboard alert: {Type} - {Title}", alert.Type, alert.Title);
                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to create dashboard alert: {Type}", alert.Type);
                return false;
            }
        }
    }

    // SignalR Hub for real-time dashboard communication
    public class DashboardHub(IRealtimeDashboardService dashboardService, ILogger<DashboardHub> logger) : Hub
    {
        private readonly IRealtimeDashboardService _dashboardService = dashboardService;
        private readonly ILogger<DashboardHub> _logger = logger;

        public async Task SubscribeToUpdates()
        {
            await _dashboardService.SubscribeToUpdatesAsync(Context.ConnectionId);
        }

        public async Task UnsubscribeFromUpdates()
        {
            await _dashboardService.UnsubscribeFromUpdatesAsync(Context.ConnectionId);
        }

        public override async Task OnDisconnectedAsync(Exception? exception)
        {
            await _dashboardService.UnsubscribeFromUpdatesAsync(Context.ConnectionId);
            await base.OnDisconnectedAsync(exception);
        }
    }
}

File: 5_WebApps/ShopERP/Services/RealtimeDashboardService.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, Subject, BehaviorSubject } from 'rxjs';
import { HubConnection, HubConnectionBuilder, LogLevel } from '@microsoft/signalr';
import { Order, OrderMetrics } from './order-management.service';

export interface DashboardAlert {
    id: string;
    type: AlertType;
    title: string;
    message: string;
    severity: AlertSeverity;
    createdAt: Date;
    expiresAt?: Date;
    isRead: boolean;
    actionUrl?: string;
    metadata: Record<string, any>;
}

export enum AlertType {
    OrderCreated = 'orderCreated',
    OrderUpdated = 'orderUpdated',
    OrderCancelled = 'orderCancelled',
    HighValueOrder = 'highValueOrder',
    SystemError = 'systemError',
    LowInventory = 'lowInventory',
    StaffAssignment = 'staffAssignment'
}

export enum AlertSeverity {
    Info = 'info',
    Warning = 'warning',
    Error = 'error',
    Critical = 'critical'
}

@Injectable({
    providedIn: 'root'
})
export class RealtimeDashboardService {
    private hubConnection: HubConnection | null = null;
    private readonly apiUrl = '/api/dashboard';
    private readonly hubUrl = '/hubs/order';
    
    // Observables for real-time updates
    private orderUpdateSubject = new Subject<Order>();
    private metricsUpdateSubject = new BehaviorSubject<OrderMetrics | null>(null);
    private alertSubject = new Subject<DashboardAlert>();
    
    // Public observables
    public orderUpdates$ = this.orderUpdateSubject.asObservable();
    public metricsUpdates$ = this.metricsUpdateSubject.asObservable();
    public alerts$ = this.alertSubject.asObservable();
    
    private subscribedConnections = new Set<string>();
    
    constructor(private http: HttpClient) {}

    async getCurrentMetrics(): Promise<OrderMetrics> {
        try {
            const response = await this.http.get<OrderMetrics>(`${this.apiUrl}/metrics`).toPromise();
            if (response) {
                this.metricsUpdateSubject.next(response);
            }
            return response || this.getDefaultMetrics();
        } catch (error) {
            console.error('Failed to get current dashboard metrics:', error);
            return this.getDefaultMetrics();
        }
    }

    async getRecentOrders(count: number = 10): Promise<Order[]> {
        try {
            return await this.http.get<Order[]>(`${this.apiUrl}/orders/recent?count=${count}`).toPromise() || [];
        } catch (error) {
            console.error('Failed to get recent orders:', error);
            return [];
        }
    }

    async getActiveOrders(): Promise<Order[]> {
        try {
            return await this.http.get<Order[]>(`${this.apiUrl}/orders/active`).toPromise() || [];
        } catch (error) {
            console.error('Failed to get active orders:', error);
            return [];
        }
    }

    async broadcastOrderUpdate(order: Order): Promise<boolean> {
        try {
            if (this.subscribedConnections.size === 0) {
                return false; // No subscribers
            }

            // Invalidate relevant caches (handled by server)
            
            // Create alert for high-value orders
            if (order.totalAmount > 1000000) { // 1 million VND
                await this.createAlert({
                    type: AlertType.HighValueOrder,
                    title: 'High Value Order',
                    message: `Order ${order.id} with total ${order.totalAmount.toLocaleString()} VND`,
                    severity: AlertSeverity.Info,
                    metadata: {
                        orderId: order.id,
                        totalAmount: order.totalAmount
                    }
                });
            }

            // Broadcast to all subscribed clients (handled by SignalR hub)
            this.orderUpdateSubject.next(order);

            // Also broadcast updated metrics
            await this.broadcastMetricsUpdate();

            console.log('Broadcasted order update:', order.id);
            return true;
        } catch (error) {
            console.error('Failed to broadcast order update:', error);
            return false;
        }
    }

    async broadcastMetricsUpdate(): Promise<boolean> {
        try {
            if (this.subscribedConnections.size === 0) {
                return false; // No subscribers
            }

            const metrics = await this.getCurrentMetrics();
            this.metricsUpdateSubject.next(metrics);

            console.log('Broadcasted metrics update');
            return true;
        } catch (error) {
            console.error('Failed to broadcast metrics update:', error);
            return false;
        }
    }

    async subscribeToUpdates(connectionId: string): Promise<boolean> {
        try {
            this.subscribedConnections.add(connectionId);

            // Send initial data
            const metrics = await this.getCurrentMetrics();
            const recentOrders = await this.getRecentOrders(10);
            const activeOrders = await this.getActiveOrders();
            const alerts = await this.getActiveAlerts();

            console.log('Client subscribed to dashboard updates:', connectionId);
            return true;
        } catch (error) {
            console.error('Failed to subscribe client to updates:', connectionId, error);
            return false;
        }
    }

    async unsubscribeFromUpdates(connectionId: string): Promise<boolean> {
        try {
            this.subscribedConnections.delete(connectionId);
            console.log('Client unsubscribed from dashboard updates:', connectionId);
            return true;
        } catch (error) {
            console.error('Failed to unsubscribe client from updates:', connectionId, error);
            return false;
        }
    }

    async getActiveAlerts(): Promise<DashboardAlert[]> {
        try {
            return await this.http.get<DashboardAlert[]>(`${this.apiUrl}/alerts`).toPromise() || [];
        } catch (error) {
            console.error('Failed to get active alerts:', error);
            return [];
        }
    }

    async createAlert(alert: Partial<DashboardAlert>): Promise<boolean> {
        try {
            const fullAlert: DashboardAlert = {
                id: this.generateGuid(),
                type: alert.type || AlertType.SystemError,
                title: alert.title || '',
                message: alert.message || '',
                severity: alert.severity || AlertSeverity.Info,
                createdAt: new Date(),
                expiresAt: alert.expiresAt,
                isRead: false,
                actionUrl: alert.actionUrl,
                metadata: alert.metadata || {}
            };

            // Broadcast alert to all subscribed clients
            if (this.subscribedConnections.size > 0) {
                this.alertSubject.next(fullAlert);
            }

            console.log('Created dashboard alert:', alert.type);
            return true;
        } catch (error) {
            console.error('Failed to create dashboard alert:', alert.type, error);
            return false;
        }
    }

    // SignalR connection management
    async startConnection(): Promise<void> {
        try {
            this.hubConnection = new HubConnectionBuilder()
                .withUrl(this.hubUrl)
                .withAutomaticReconnect()
                .configureLogging(LogLevel.Information)
                .build();

            this.hubConnection.on('OrderUpdated', (order) => {
                this.orderUpdateSubject.next(order);
            });

            this.hubConnection.on('MetricsUpdated', (metrics) => {
                this.metricsUpdateSubject.next(metrics);
            });

            this.hubConnection.on('AlertCreated', (alert) => {
                this.alertSubject.next(alert);
            });

            await this.hubConnection.start();
            console.log('SignalR connection started');
        } catch (error) {
            console.error('Failed to start SignalR connection:', error);
        }
    }

    async stopConnection(): Promise<void> {
        try {
            if (this.hubConnection) {
                await this.hubConnection.stop();
                this.hubConnection = null;
                console.log('SignalR connection stopped');
            }
        } catch (error) {
            console.error('Failed to stop SignalR connection:', error);
        }
    }

    private getDefaultMetrics(): OrderMetrics {
        return {
            totalOrders: 0,
            pendingOrders: 0,
            processingOrders: 0,
            completedOrders: 0,
            cancelledOrders: 0,
            totalRevenue: 0,
            averageOrderValue: 0,
            ordersPerHour: 0,
            revenuePerHour: 0,
            statusBreakdown: []
        };
    }

    private generateGuid(): string {
        return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
            const r = Math.random() * 16 | 0;
            const v = c === 'x' ? r : (r & 0x3 | 0x8);
            return v.toString(16);
        });
    }
}
